### PR TITLE
Cryptographic Sortition for Proof of Stake

### PR DESCRIPTION
**Pull request by rmnblm**
Good morning Bazo'ers 👋 

As described in my paper "[Cryptographic Sortition for PoS in Bazo](https://github.com/rmnblm/papers/blob/master/cryptographic_sortition/README.md)", published in the early spring 2018, I changed the Proof of Stake algorithm from using a local seed file to an RSA public-private keypair. 

The adjusted PoS condition now looks as follows:

screenshot 2018-10-29 at 08 42 55

The main two steps which have changed from the old *Seed*-version to the new *RSA*-version now looks as follows:

**Step 1: CommitmentKey** 
A user publishes a StakeTx in order to join the pool of validators. Now, the StakeTx contains the RSA public key `CommitmentKey` instead of a hashed seed.

screenshot 2018-10-29 at 08 34 20

**Step 2: CommitmentProof**
When a validator fulfils the PoS condition, the validator publishes the block containing a `CommitmentProof`.

screenshot 2018-10-29 at 08 33 34

Have a nice day y'all 🍷 

Roman

### CODE REVIEW

File: protocol/block.go [hunk 1]
```diff
@@ -4,14 +4,14 @@ import (
...
-
 	"github.com/willf/bloom"
 	"golang.org/x/crypto/sha3"
 )
 
 const (
 	HASH_LEN                = 32
-	MIN_BLOCKSIZE           = 318
+	HEIGHT_LEN				= 4
+	MIN_BLOCKSIZE           = 510
```
Comment (tbocek) [inline protocol/block.go]: How did you calculate 510?

File: main.go [hunk 2]
```diff
@@ -4,38 +4,45 @@ import (
...
 
-	multisigPubKey, _, err := storage.ExtractKeyFromFile(multisigFileName)
+	multisigPubKey, _, err := storage.ExtractECDSAKeyFromFile(multisigFileName)
 	if err != nil {
 		logger.Printf("%v\n", err)
 		return
 	}
 
-	miner.Init(&validatorPubKey, &multisigPubKey, seedFileName)
+	commPrivKey, err := protocol.ExtractRSAKeyFromFile(commFileName)
```
Comment (tbocek) [inline main.go]: Is it easy to create one if the file does not exist?

**Review by tbocek — COMMENTED** [review]
some questions

File: main.go (hunk 2 above)
Comment (rmnblm) [reply main.go]: The bazo-miner automatically creates a file, generates an RSA keypair and writes it into the new file.

File: protocol/block.go (hunk 1 above)
Comment (rmnblm) [reply protocol/block.go]: By deleting `Seed` and `HashedSeed` from the block, the size of the block decreases by 

`MIN_BLOCKSIZE - len(Seed) - len(HashedSeed) = 318 - 32 - 32 = 254`

By adding the `CommitmentProof` to the block, the size of the block then decreases by

`254 + len(CommitmentProof) = 254 + 256 = 510`

Note that this is a constant value and if the constant value of `COMM_KEY_LENGTH` changes, the `MIN_BLOCKSIZE` constant has to be changed too.

**Comment by mchetelat** [comment]
What about existing staking accounts working dependent on their seed file?

**Comment by tbocek** [comment]
They need to switch to the new chain. There is no upgrade path. That means we need to notify the ones running nodes and tell them to update

**Review by mchetelat — APPROVED** [review]
Nice work. Thanks a lot @rmnblm !

**Pull request merged**